lib/SILOptimizer/Mandatory/AddressLowering.cpp

```diff
@@ -3392,6 +3399,23 @@ static void emitEndBorrows(SILValue value, AddressLoweringState &pass) {
       });
 }
 
+/// Emit end_borrows at the lifetime ends of the guaranteed value which
+/// encloses \p enclosingValue.
+static void
+emitEndBorrowsAtEnclosingGuaranteedBoundary(SILValue lifetimeToEnd,
+                                            SILValue enclosingValue,
+                                            AddressLoweringState &pass) {
+  visitBorrowIntroducers(enclosingValue, [&](SILValue introducer) {
+    auto borrow = BorrowedValue(introducer);
+    borrow.visitLocalScopeEndingUses([&](Operand *use) {
```

Comment:
Does this insert end_borrows correctly when the BorrowedValue has reborrows  ?

Reply:
Thanks for raising that issue!  It seems, however, like this shouldn't actually be possible:

For it to happen, we'd need to be rewriting a guaranteed opaque value `v` which has some projection (i.e. `struct|tuple_extract`/`destructure_struct|tuple`) user.  Some introducers are guaranteeing that opaque value.  Because it's not legal to change representation of a guaranteed opaque value via forwarding (`SIL.rst:Forwarding Address-Only Values`), each introducer `i` must also be an opaque value (`i`'s type is some aggregate in which `v`'s type appears).  In order to have a reborrow of an introducer, we'd need a guaranteed phi of that introducer's type.  In other words, we'd need a guaranteed phi of opaque type.  But that's not allowed:
```
        // Address-only values are potentially unmovable when borrowed. See also
        // checkOwnershipForwardingInst. A phi implies a move of its arguments
        // because they can't necessarily all reuse the same storage.
        require((!arg->getType().isAddressOnly(F)
                 || arg->getOwnershipKind() != OwnershipKind::Guaranteed),
                "Guaranteed address-only phi not allowed--implies a copy");
```

This is a pretty subtle point so I added a comment and some assertions.

Reply:
Thanks!

Reply:
Thanks. I think you're also assuming a single introducer now. So you may want to check that.

Reply:
Considering that `struct` and `tuple` both change representation, this isn't an _assumption_, right?  If so, it's still definitely something else to assert.  Or are there other ways we could have multiple introducers?

Reply:
Collected introducers into a TinyPtrVector and asserted the size was 1.

Reply:
So, if you're asserting there's only one introducer, then why is there a loop? Why can't you just have a single optional introducer and assert that it isn't set twice?

Reply:
Sure, that would be fine too.

Reply:
https://github.com/apple/swift/pull/67443